org.eclipse.xtext/src/org/eclipse/xtext/xtext/RuleNames.java

```diff
@@ -146,13 +146,15 @@ public RuleNames(Grammar grammar, boolean installAdapter) {
 			String uniqueName = name;
 			String antlrRuleName;
 			if (names.containsKey(name)) {
+				names.put(name, rule);
```

Comment:
I do not understand the fix. Can you please provide a test scenario, that illustrates the problem?

Reply:
In fact this looks wrong to me, since we put a second rule instance into the map by simple name. But the rule from the super grammar should not replace the rule from the sub-grammer (which is already in the map)

Reply:
It started with the following stacktrace:
```
Caused by: java.lang.IllegalStateException: AlgorithmInterface
	at org.eclipse.xtext.xtext.RuleNames.<init>(RuleNames.java:157)
	at org.eclipse.xtext.xtext.RuleNames.getRuleNames(RuleNames.java:94)
	at org.eclipse.xtext.xtext.generator.XtextGeneratorLanguage.initialize(XtextGeneratorLanguage.java:296)
	at org.eclipse.xtext.xtext.generator.XtextGeneratorLanguage.initialize(XtextGeneratorLanguage.java:283)
	at org.eclipse.xtext.xtext.generator.StandardLanguage.initialize(StandardLanguage.java:189)
	at org.eclipse.xtext.xtext.generator.XtextGenerator.initialize(XtextGenerator.java:185)
	at org.eclipse.xtext.xtext.generator.XtextGenerator.checkConfigurationInternal(XtextGenerator.java:139)
	at org.eclipse.emf.mwe.core.lib.AbstractWorkflowComponent2.checkConfiguration(AbstractWorkflowComponent2.java:21)
	at org.eclipse.emf.mwe.core.lib.Mwe2Bridge.preInvoke(Mwe2Bridge.java:65)
	at org.eclipse.emf.mwe.core.lib.AbstractWorkflowComponent.preInvoke(AbstractWorkflowComponent.java:197)
	at org.eclipse.emf.mwe2.runtime.workflow.AbstractCompositeWorkflowComponent.preInvoke(AbstractCompositeWorkflowComponent.java:29)
	at org.eclipse.emf.mwe2.runtime.workflow.Workflow.run(Workflow.java:18)
	at org.eclipse.emf.mwe2.launch.runtime.Mwe2Runner.run(Mwe2Runner.java:102)
	... 4 more
```

The debugger showed that the value of `uniqueName` here is not qualified which lead to a problem later in the block:

```
if (!usedUniqueNames.add(uniqueName)) {
	throw new IllegalStateException(uniqueName);
}
```

So I thought the rule was meant to be overwritten in names.

Unfortunately the grammars are large, and I cannot share them. I'll try to figure out a minimal set that shows the problem.